[WIP] Add DmgWifiScheduler class which handles the scheduling features for 80211ad

tommyazz opened the pull request:
This class provides a set of methods that handle the reception of ADDTS requests, the admission policy for received requests and the organization of the DTI.

tommyazz committed:
Update gitignore

tommyazz committed:
Create class DmgWifiScheduler

tommyazz committed:
Add class description for DmgWifiScheduler

tommyazz committed:
Add custom dmg error model to lroom scenario

tommyazz committed:
Add interface between DmgApWifiMac and DmgWifiScheduler

Add method SetScheduler in DmgApWifiMac class and SetMac in DmgWifiScheduler.

tommyazz committed:
Setup callback to handle a received ADDTS request at DmgScheduler

tommyazz committed:
Connect BI start and DTI start traces to DmgWifiScheduler

tommyazz committed:
Fix indentation issues

tommyazz committed:
Move methods of type Add Allocation from DmgApWifiMac to DmgScheduler

All methods performing the allocation of SPs or CBAPs are moved from the DmgApWifiMac to the DmgWifiScheduler. These methods still exist, but not implement, in DmgApWifiMac to temporarily preserve compatibility with old examples.

tommyazz committed:
Move old evaluate scripts in examples/wireless/80211ad

Examples are disabled by default in ns3-80211ad due to old compatibility issue. Therefore, we should not have any problem.

tommyazz committed:
All methods for allocating an SP or CBAP have been moved to DmgWifiSched

tommyazz committed:
Create callback to handle a DELTS request at DmgWifiScheduler

tommyazz committed:
Type definition for AllocationFieldList iterator and const iterator

tommyazz committed:
Add methods to store and manage the ADDTS requests received

tommyazz committed:
Fix problem with connection to Traces

DmgWifiScheduler was created in the wrong place at DmgApWifiMac in order to connect traces.

tommyazz committed:
Add map to store the allocated ADDTS requests

m_allocatedAddtsRequests stores the ADDTS requests that have been accepted and allocated by the PCP/AP.
The original allocation parameters for each request  are stored in this map.

tommyazz committed:
Add two methods to handle an ADDTS request

The ADDTS request is handled differently whether or not the allocation already exists. 
ModifyExistingAllocation handles the acceptance or rejection when the AP receives an ADDTS request related to an existing allocation. Whereas, AddNewAllocation handles the acceptance or rejection  of an ADDTS request NOT related to an existing one.

tommyazz committed:
Add function  to get the allocation duration (between min and max))

creating the admission policy for new ADDTS request at AddNewAllocation

tommyazz committed:
Fix isConnected problem in connection to traces at DmgWifiScheduler

tommyazz committed:
Fix bug in ModifyAllocation of DmgWifiScheduler

The copy of an object was modified instead of the object itself.

tommyazz committed:
Allocations are marked as announced when the DTI at the PCP/AP starts 

In this case, the DTI is not entirely allocated as CBAP.

tommyazz committed:
Develop methods AddNewAllocation and ModifyExistingAllocation

Introduced two separated allocation lists:
m_addtsAllocationList stores all the allocations granted by the AP through an ADDTS request.
m_globalAllocationList is the one sent in an DMG beacon frame. It adds to  m_addtsAllocationList a series of broadcast CBAPs where there is DTI time not allocated. The function which performing this feature is going to be implemented in later commits

tommyazz committed:
Create method to place broadcast CBAPs among requested allocations

tommyazz committed:
Introduced m_interAllocationDistance parameter

This parameter represents the time distance between to allocations that is used as broadcast CBAP.

tommyazz committed:
Add method to handle the allocations in the list when one is removed

tommyazz committed:
Modify AddBroadcastCbapAllocations based on the interAllocation distance

tommyazz committed:
Create ADDTS response rx traced callback at DmgStaWifiMac

tommyazz committed:
Calling StartBeaconInterval from dmgWifiScheduler

The function calculateBTIvariables of DmgApWifiMac is scheduled when the BI ends. Together with the evaluation of ADDTS request at DmgWifiScheduler. It happens that the former method is called before evaluating the requests (tough at the same simulation time) resulting in an empy extended schedule element that impacts the right calculation of the DMG beacon duration.

tommyazz committed:
Create simple ADDTS request for an SP in evaluate-lroom

tommyazz committed:
Add guard time between adjacent allocations

tommyazz committed:
Moved CalculationBTIVariables at the start of beacon interval

The BHI duration at DmgWifiScheduler differed from the one at DmgAp, because the above method was called at the start of the BTI. This method calculates the duration of the BTI based on duration of a DMG beacon.

tommyazz committed:
Temporary Fix to issue 9. Create dummy ESE with MAX size

tommyazz committed:
Add some logs in DmgWifiScheduler class

mattia-lecci commented on src/wifi/model/status-code.h:
```diff
@@ -30,6 +30,7 @@ enum StatusCodeValue
   STATUS_CODE_SUCCESS = 0,
   STATUS_CODE_FAILURE = 1,
   STATUS_CODE_REJECTED_WITH_SUGGESTED_CHANGES = 39,
+  STATUS_CODE_REJECTED_FOR_DELAY_PERIOD = 47,
```
reference?

mattia-lecci commented on src/wifi/model/status-code.cc:
```diff
@@ -44,6 +44,12 @@ StatusCode::IsSuccess (void) const
   return (m_code == 0);
 }
 
+void
+StatusCode::SetStatusCodeValue (StatusCodeValue value)
+{
+  m_code = value;
+}
+
```
maybe a non-empty constructor could be useful?

tommyazz replied:
I just implemented a function that was declared in the header file but not actually implemented.
I think the usage of this element is clearer in this way.

mattia-lecci commented on scratch/evaluate_qd_channel_lroom_scenario.cc:
```diff
@@ -386,6 +414,9 @@ main (int argc, char *argv[])
   /* Sensitivity model includes implementation loss and noise figure */
   spectrumWifiPhy.Set ("CcaMode1Threshold", DoubleValue (-79));
   spectrumWifiPhy.Set ("EnergyDetectionThreshold", DoubleValue (-79 + 3));
+  /* Custom error rate model for IEEE 802.11ad */
+  spectrumWifiPhy.SetErrorRateModel ("ns3::DmgErrorModel",
+                                     "FileName", StringValue ("DmgFiles/ErrorModel/LookupTable_1458.txt"));
```
Looking at DmgWifiPhyHelper, the default ErrorRateModel is ns3::SensitivityModel60GHz. Is this not correct? Why should one be chosen in place of the other?

mattia-lecci replied:
Also, the same should be applied to all scenarios (if correct).
If this is the correct model to be used by default, I would act on DmgWifiPhyHelper itself rather than from each scratch script

tommyazz replied:
This error rate model comes with the QD channel model. And it has been used in all QD evaluation scripts except this one. Therefore I added it. I am not able to say why we should use it since I didn't work on it. But in principle, this error model should be for accurate than sensitivity. I would retain sensitivity as the default one for now.

mattia-lecci commented on scratch/evaluate_qd_channel_lroom_scenario.cc:
```diff
@@ -512,6 +543,10 @@ main (int argc, char *argv[])
+  /* Create DMG Scheduler and setup */
+  dmgScheduler = CreateObject<DmgWifiScheduler> ();
+  apWifiMac->SetScheduler (dmgScheduler);
+
```
this should be done, instead, in DmgWifiHelper, and extracted just like all other components (see lines 537-544)

tommyazz replied:
WIP

mattia-lecci commented on src/wifi/model/dmg-sta-wifi-mac.cc:
```diff
@@ -1203,7 +1206,7 @@ DmgStaWifiMac::StartDataTransmissionInterval (void)
         {
           AllocationField field;
-          for (AllocationFieldList::iterator iter = m_allocationList.begin (); iter != m_allocationList.end (); iter++)
+          for (AllocationFieldListI iter = m_allocationList.begin (); iter != m_allocationList.end (); iter++)
```
for brevity "auto" could also have been used

mattia-lecci commented on src/wifi/model/dmg-sta-wifi-mac.cc:
```diff
@@ -199,6 +199,9 @@ DmgStaWifiMac::GetTypeId (void)
                      "The BeamLink maintenance timer associated to a link has expired.",
                      MakeTraceSourceAccessor (&DmgStaWifiMac::m_beamLinkMaintenanceTimerExpired),
                      "ns3::DmgStaWifiMac::BeamLinkMaintenanceTimerExpiredTracedCallback")
+    .AddTraceSource ("ADDTSResponse", "Received an ADDTS response from PCP/AP.",
```
The [ns-3 coding style](https://www.nsnam.org/develop/contributing-code/coding-style/#Name_encoding) explicitly says "Do not use all capital letters, even for acronyms such as EDCA: use Edca instead."
Thus, I would suggest to change this line as:
.AddTraceSource ("AddtsResponse", "Received an ADDTS Response from PCP/AP.",

tommyazz replied:
I adopted the ns3 coding style in 99,99 % of the code I've written. I just wanted to adopt the same style as the other callbacks declared by Hany. We would need to change name to many of them then.

mattia-lecci commented on src/wifi/model/dmg-sta-wifi-mac.h:
```diff
@@ -629,6 +629,14 @@ class DmgStaWifiMac : public DmgWifiMac
+   * TracedCallback signature for an ADDTS response received.
+   *
+   * \param address The MAC address of the station.
+   * \param transmissionLink The new transmission link.
```
It was probably removed. Now there are status and dmgTspec, instead

tommyazz replied:
I don't understand, I added this traced callback.

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+ * This class provides the implementation of scheduling features related to
+ * IEEE 802.11ad. In particular, this class organizes the medium access 
+ * according to the availability of contention-free access periods (SPs)
+ * and contention-based access periods (CBAPs) as foresee by 802.11ad amendment.
```
foresee -> provided

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   */
+  virtual void ManageAddtsRequests (void);
+  /**
+   * Implement the policy that accept or reject a new ADDTS request.
```
accepts, rejects

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  /**
+   * Update start time and remaining DTI time for the next request to be evaluated.
+   */
+  virtual void UpdateStartandRemainingTime (void);
```
and -> And

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param iter The iterator pointing to the next element in the addtsAllocationList.
+   * \param removedAlloc The removed allocation information.
+   */
+  virtual void AdjustExistingAllocations (AllocationFieldListI iter, AllocationField removedAlloc);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param address The MAC address of the source STA.
+   * \param element The Dmg Tspec Element associated with the request.
+   */
+  virtual void ReceiveAddtsRequest (Mac48Address address, DmgTspecElement element);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param address The MAC address of the requesting STA.
+   * \param info The Dmg Allocation Info field associated with the request.
+   */
+  virtual void ReceiveDeltsRequest (Mac48Address address, DmgAllocationInfo info);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  /**
+   * Manage the ADDTS requests received at the PCP/AP during the last DTI.
+   */
+  virtual void ManageAddtsRequests (void);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param info The DMG Allocation Info element of the request.
+   * \return The Status Code to be included in the ADDTS response.
+   */
+  virtual StatusCode AddNewAllocation (uint8_t sourceAid, DmgTspecElement dmgTspec, DmgAllocationInfo info);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param info The DMG Allocation Info element of the request.
+   * \return The Status Code to be included in the ADDTS response.
+   */
+  virtual StatusCode ModifyExistingAllocation (uint8_t sourceAid, DmgTspecElement dmgTspec, DmgAllocationInfo info);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param maxAllocation The desired allocation in us for each allocation period.
+   * \return The allocation duration for the allocation period.
+   */
+  virtual uint32_t GetAllocationDuration (uint32_t minAllocation, uint32_t maxAllocation);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  /**
+   * Add broadcast CBAP allocations in the DTI.
+   */
+  virtual void AddBroadcastCbapAllocations (void);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  /**
+   * \return The TS Delay element to be included in the ADDTS response.
+   */
+  virtual TsDelayElement GetTsDelayElement (void);
```
why public?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   */
+  virtual void AddBroadcastCbapAllocations (void);
+  /**
+   * \return The TS Delay element to be included in the ADDTS response.
```
Not clear enough

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param staticAllocation Is the allocation static.
+   * \param allocationStart The start time of the allocation relative to the beginning of DTI.
+   * \param blockDuration The duration of the allocation period.
+   * \return The start of the next allocation period.
```
unit of measure?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param destAid The AID of the destination DMG STA.
+   * \param allocationStart The start time of the allocation relative to the beginning of DTI.
+   * \param blockDuration The duration of the allocation period.
+   * \return The start of the next allocation period.
```
unit of measure?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  /**
+   * Allocate CBAP period to be announced in DMG Beacon or Announce Frame.
+   * \param staticAllocation Is the allocation static.
+   * \param allocationStart The start time of the allocation relative to the beginning of DTI.
```
unit of measure?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * Allocate CBAP period to be announced in DMG Beacon or Announce Frame.
+   * \param staticAllocation Is the allocation static.
+   * \param allocationStart The start time of the allocation relative to the beginning of DTI.
+   * \param blockDuration The duration of the allocation period.
```
unit of measure?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param sourceAid The AID of the source DMG STA.
+   * \param destAid The AID of the destination DMG STA.
+   */
+  void AllocateDTIAsServicePeriod (AllocationID allocationId, uint8_t sourceAid, uint8_t destAid);
```
DTI -> Dti
(again for ns-3 codestyle)

tommyazz replied:
This is from Hany tough

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  /**
+   * \return The full extended schedule element.
+   */
+  Ptr<ExtendedScheduleElement> GetFullExtendedScheduleElement (void);
```
why "full"?

tommyazz replied:
it is a dummy ESE with the maximum number of allocations it can carry. This is related to issue #9

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  Ptr<DmgApWifiMac> m_mac;                     //!< Pointer to the MAC high of PCP/AP.
+
+  /* Access Period Allocations */
+  AllocationFieldList m_globalAllocationList;  //!< Global list of access period allocations in DTI.
```
why global? possibly misleading name

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  /**
+   * Creates a dummy ESE with maximum size.
+   */
+  void CreateFullExtentdedScheduleElement (void);
```
why full?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+   * \param address The MAC address of the PCP/AP.
+   * \param dtiDuration The duration of the current DTI.
+   */
+  void DataTransferIntervalStarted (Mac48Address address, Time dtiDuration);
```
I'd prefer the shortened versions BiStarted, DtiStarted, etc. but it's just a personal preference :)

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  Time m_dtiDuration;                          //!< The length of the DTI period.
+  Time m_biStartTime;                          //!< The start time of the BI Interval.
+  Time m_atiStartTime;                         //!< The start time of the ATI Interval.
+  Time m_dtiStartTime;                         //!< The start time of the DTI Interval.
```
maybe just m_currentAccessPeriodStartTime and m_currentAccessPeriodDuration will suffice combined with m_accessPeriod?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  void AddBroadcastCbap (void);
+
+  /* Channel Access Period */
+  ChannelAccessPeriod m_accessPeriod;          //!< The type of the current channel access period.
```
I'd prefer m_currentAccessPeriod

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  } AddtsRequest;
+  typedef std::vector<AddtsRequest> AddtsRequestList;
+  typedef AddtsRequestList::const_iterator AddtsRequestListCI;
+  AddtsRequestList m_receiveAddtsRequests;     //!< The list containing the ADDTS requests received during the DTI.
```
*received

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  typedef AddtsRequestList::const_iterator AddtsRequestListCI;
+  AddtsRequestList m_receiveAddtsRequests;     //!< The list containing the ADDTS requests received during the DTI.
+
+  /* An allocation is uniquely identified by the tuple: Allocation Id, Source Aid, Destination Aid (802.11ad 10.4). */
```
*ID, *AID

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  AllocatedRequestMap m_allocatedAddtsRequests; //!< The map containing the allocated ADDTS requests with their original allocation parameters.
+
+  uint32_t m_minBroadcastCbapDuration;          //!< The minimum duration of a broadcast CBAP to be present in the DTI.
+  uint32_t m_interAllocationDistance;           //!< The distance between two allocations to be used as broadcast CBAP.
```
so you mean interSpDistance?

tommyazz replied:
Yes, but a requested allocation can also be a CBAP

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  } NextAllocationTiming;
+  NextAllocationTiming m_schedulingTime;        //!< Store the next start time and remaining DTI time.
+  uint32_t m_guardTime;                         //!< The guard time between allocations in us.
+  Ptr<ExtendedScheduleElement> m_fullEse;       //!< Full ESE for calculation of BTI duration.
```
why full?

tommyazz replied:
issue #9

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+      .SetGroupName ("Wifi")
+      .AddConstructor<DmgWifiScheduler> ()
+
+      .AddAttribute ("MinBroadcastCbapDuration", "The minimum duration of a broadcast CBAP in the DTI",
```
unit of measure?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+                     UintegerValue (4096),
+                     MakeUintegerAccessor (&DmgWifiScheduler::m_minBroadcastCbapDuration),
+                     MakeUintegerChecker<uint32_t> ())
+      .AddAttribute ("InterAllocationDistance", "The time distance between two adjacent allocations "
```
unit of measure?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+  m_isNonStatic = false;
+  m_isDeltsReceived = false;
+  m_guardTime = GUARD_TIME.GetMicroSeconds ();
+}
```
It's better to use [initializer lists](https://stackoverflow.com/questions/5816218/difference-between-initializer-and-default-initializer-list-in-c/5816311) instead

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+  NS_ASSERT_MSG (isConnected, "Connection to Trace DELTSReceived failed.");
+}
+
+void
+DmgWifiScheduler::CreateFullExtentdedScheduleElement (void)
+{
+  m_fullEse = Create<ExtendedScheduleElement> ();
+  uint16_t maxLength = 17;
+  AllocationField field;
+  for (uint16_t i = 0; i < maxLength; ++i)
+  {
+    m_fullEse->AddAllocationField (field);
+  }
+}
```
need explanation

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+DmgWifiScheduler::GetTsDelayElement (void)
+{
+  TsDelayElement element;
+  element.SetDelay (1);
```
i assume this is just temporary?

tommyazz replied:
Yes

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+  /* Update start and remaining DTI times
+   * This update here takes into account the possible cleanup of non-static allocations
+   * and the possible reception of DELTS requests
+   */
```
Too detailed. It should be in the documentation of the method, instead

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+    {
+      /* At least one ADDTS request has been received */
+      ManageAddtsRequests ();
+      /* Start and remaining DTI times are updated according to the allocated requests */
```
is this referred to ManageAddtsRequests ()?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+      ManageAddtsRequests ();
+      /* Start and remaining DTI times are updated according to the allocated requests */
+    }
+  AddBroadcastCbap ();
```
Why is this in BeacondIntervalEnded?

tommyazz replied:
By design, now, we evaluate all the received ADDTS requests at the end of the DTI.

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+      /* Start and remaining DTI times are updated according to the allocated requests */
+    }
+  AddBroadcastCbap ();
+  m_mac->StartBeaconInterval ();
```
isn't m_mac self-sufficient to this regard? why should it depend on the scheduler for this?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+void
+DmgWifiScheduler::ReceiveAddtsRequest (Mac48Address address, DmgTspecElement element)
+{
+  NS_LOG_INFO ("Receive ADDTS request from " << address);
```
*Received

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+      case STATUS_CODE_DENIED_DUE_TO_SPECTRUM_MANAGEMENT:
+        break;
+      default: 
+        NS_FATAL_ERROR ("ADDTS response status code not supported");
```
NS_FATAL_ERROR ("ADDTS response status code=" << status.GetStatusCodeValue () << " not supported");

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+  else
+    {
+      /* The allocation does not exist */
+      NS_LOG_DEBUG ("Cannot find the allocation");
```
maybe an assert is more appropriate?

tommyazz replied:
I'd put a WARN or (not sure) a FATAL ERROR

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+  typedef std::tuple<AllocationID, uint8_t, uint8_t> UniqueIdentifier;
+  typedef std::map<UniqueIdentifier, AddtsRequest> AllocatedRequestMap;
+  typedef AllocatedRequestMap::iterator AllocatedRequestMapI;
+  AllocatedRequestMap m_allocatedAddtsRequests; //!< The map containing the allocated ADDTS requests with their original allocation parameters.
```
maybe it would be better to call it m_allocatedAddtsRequestMap

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+    Mac48Address sourceAddr;
+    DmgTspecElement dmgTspec;
+  } AddtsRequest;
+  typedef std::vector<AddtsRequest> AddtsRequestList;
```
why not a std::map<UniqueIdentifier, AddtsRequest> instead?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+{
+  /* Manage the ADDTS requests received in the last DTI.
+   * Channel access organization during the DTI.
+   */
```
should be in the doc instead

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+            {
+              ++iter;
+            }
+        }
```
it woud be much simpler (and efficient) if m_addtsAllocationList were a map

tommyazz replied:
mmm, not sure about the efficiency of the insert operation tough. Anyhow, Hany decided that the allocation list is of type AllocationFieldList. If we want to treat the allocation list as a map we would need to change some code in other classes. Or we would need to return an AllocationFieldList from DmgWifiScheduler (i.e. copy each element in the map to a vector)

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+  StatusCode status;
+  UniqueIdentifier allocIdentifier;
+  AllocatedRequestMapI it;
+  /* Cycle over the list of received ADDTS requests; remainingDtiTime is updated each time an allocation is accepted.
```
*Loop over

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+}
+
+void
+DmgWifiScheduler::ManageAddtsRequests (void)
```
Need to discuss about this method: much easier to understand while talking

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+}
+
+void
+DmgWifiScheduler::UpdateStartandRemainingTime (void)
```
not sure about what it does

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+}
+
+void
+DmgWifiScheduler::AdjustExistingAllocations (AllocationFieldListI iter, AllocationField removedAlloc)
```
not sure about its purpose

MatteoDrago replied:
I think it is meaningful to discuss this method. My view:

- could be useful to aggregate free time, for allocating longer SPs to new users requesting resources 
- could be also useful to update the already allocated SPs, if the initial strategy was conservative and allocated less time due to high demands for multiple users
- could we mess up with the original periodicity, or is it shift-invariant? 
- could this result in additional overhead, as the PCP/AP needs to announce all these changes? 

Let's discuss

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+{
+  NS_LOG_FUNCTION (this);
+  /* Implementation of an admission policy for newly received requests. */
+  if (dmgTspec.GetAllocationPeriod () != 0)
```
allocation period == 0 is reserved. what did you want to check here?

tommyazz replied:
it's not reserved if bit15 is 1. It indicates the allocation is not periodic

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+    }
+  else
+    {
+      NS_LOG_WARN ("Allocation Format not supported");
```
FATAL_ERROR

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+      .AddConstructor<DmgWifiScheduler> ()
+
+      .AddAttribute ("MinBroadcastCbapDuration", "The minimum duration of a broadcast CBAP in the DTI",
+                     UintegerValue (4096),
```
isn't this defined by the standard?

tommyazz replied:
Nope

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,330 @@
+    uint32_t startTime;
+    uint32_t remainingDtiTime;
+  } NextAllocationTiming;
+  NextAllocationTiming m_schedulingTime;        //!< Store the next start time and remaining DTI time.
```
is it necessary to have a struct?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+}
+
+StatusCode
+DmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, DmgTspecElement dmgTspec, DmgAllocationInfo info)
```
this is an implementation of a "real" scheduler and thus not needed by its children.
I would keep this class as abstract and create a child scheduler implementing this method

tommyazz replied:
That's the idea in the future. I will do this before merging the PR (among the last things to do)

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+}
+
+StatusCode
+DmgWifiScheduler::ModifyExistingAllocation (uint8_t sourceAid, DmgTspecElement dmgTspec, DmgAllocationInfo info)
```
Again, i would move this to a child scheduler

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+              m_schedulingTime.remainingDtiTime -= timeDifference;
+              status.SetSuccess ();
+            }
+          goto reorderList;
```
avoid goto

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+          goto reorderList;
+        }
+      status.SetFailure ();
+      goto doNothing; // The request is not accepted; maintaining old allocation duration
```
avoid goto

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+    return status;
+  doNothing:
+    /* no need to update next start time */ 
+    return status;
```
avoid goto

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+      m_isAllocationModified = false;
+      m_isNonStatic = false;
+      m_isDeltsReceived = false;
+    }  
```
unclear. what is the meaning of this part?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+}
+
+void 
+DmgWifiScheduler::AddBroadcastCbapAllocations (void)
```
better to talk about this

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+DmgWifiScheduler::GetBroadcastCbapAllocation (bool staticAllocation, uint32_t allocationStart, uint32_t blockDuration)
+{
+  uint16_t numberCbapBlocks = blockDuration / MAX_CBAP_BLOCK_DURATION;
+  uint16_t lastCbapLegnth = blockDuration % MAX_CBAP_BLOCK_DURATION;
```
*lastCbapLength

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+
+AllocationFieldList
+DmgWifiScheduler::GetBroadcastCbapAllocation (bool staticAllocation, uint32_t allocationStart, uint32_t blockDuration)
+{
```
NS_LOG_FUNCTION

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+                                                 uint8_t sourceAid, uint8_t destAid, uint32_t allocationStart, uint16_t blockDuration)
+{
+  NS_LOG_FUNCTION (this);
+  return (AddAllocationPeriod (allocationId, allocationType, staticAllocation, sourceAid, destAid,
```
unnecessary brackets

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+      allocation = (*iter);
+      if (!allocation.IsPseudoStatic () && allocation.IsAllocationAnnounced ())
+        {
+          m_isNonStatic = true;
```
what is this for?

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+
+void
+DmgWifiScheduler::SetAllocationsAnnounced (void)
+{
```
NS_LOG_FUNCTION

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+
+TsDelayElement
+DmgWifiScheduler::GetTsDelayElement (void)
+{
```
NS_LOG_FUNCTION

mattia-lecci commented on src/wifi/model/dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,785 @@
+}
+
+void 
+DmgWifiScheduler::BeaconIntervalStarted (Mac48Address address, Time biDuration, Time bhiDuration, Time atiDuration)
```
NS_LOG_FUNCTION

mattia-lecci replied:
same goes for other meaningful functions

mattia-lecci commented on src/wifi/model/dmg-ap-wifi-mac.h:
```diff
@@ -11,6 +11,7 @@
 #include "amsdu-subframe-header.h"
 #include "dmg-beacon-dca.h"
 #include "dmg-wifi-mac.h"
+#include "dmg-wifi-scheduler.h"
```
i think a forward declaration should be enough

tommyazz replied:
The forward declaration is for dmgWifiScheduler with respect to DmgApWifiMac

mattia-lecci commented on src/wifi/model/dmg-ap-wifi-mac.h:
```diff
@@ -238,19 +150,14 @@ class DmgApWifiMac : public DmgWifiMac
    * \param elem The TSPEC element.
    */
-  void SendDmgAddTsResponse (Mac48Address to, StatusCode code, TsDelayElement &delayElem, DmgTspecElement &elem);
+  void SendDmgAddTsResponse (Mac48Address to, StatusCode code, TsDelayElement delayElem, DmgTspecElement elem);
```
why not passing by reference?

mattia-lecci commented on src/wifi/model/dmg-ap-wifi-mac.h:
```diff
@@ -302,11 +210,16 @@ class DmgApWifiMac : public DmgWifiMac
    */
   Ptr<DmgCapabilities> GetDmgCapabilities (void) const;
 
+  void StartBeaconInterval (void);
```
why was this moved up here?

mattia-lecci commented on src/wifi/model/dmg-ap-wifi-mac.cc:
```diff
@@ -177,6 +177,9 @@ DmgApWifiMac::GetTypeId (void)
     .AddTraceSource ("ADDTSReceived", "The PCP/AP received DMG ADDTS Request.",
                      MakeTraceSourceAccessor (&DmgApWifiMac::m_addTsRequestReceived),
                      "ns3::DmgApWifiMac::AddTsRequestReceivedTracedCallback")
+    .AddTraceSource ("DELTSReceived", "The PCP/AP received DELTS Request.",
```
DELTSReceived -> DeltsReceived
https://www.nsnam.org/develop/contributing-code/coding-style/#Name_encoding

mattia-lecci commented on src/wifi/model/dmg-ap-wifi-mac.cc:
```diff
@@ -217,6 +220,21 @@ DmgApWifiMac::DoDispose ()
+  NS_LOG_FUNCTION (this);
+  m_dmgScheduler = dmgScheduler;
+  m_dmgScheduler->SetMac (this);
+  m_dmgScheduler->Initialize ();
```
maybe this should be done by DmgWifiHelper

mattia-lecci commented on src/wifi/model/dmg-ap-wifi-mac.cc:
```diff
@@ -691,7 +556,8 @@ DmgApWifiMac::CalculateBTIVariables (void)
   if (m_scheduleElement)
     {
-      beacon.AddWifiInformationElement (GetExtendedScheduleElement ());
+      /* TEMPORARY FIX: GET A DUMMY FULL EXTENDED SCHEDULE ELEMENT */
+      beacon.AddWifiInformationElement (m_dmgScheduler->GetFullExtendedScheduleElement ());
```
we should discuss about this

tommyazz replied:
Yes (issue #9)

mattia-lecci commented on src/wifi/model/dmg-ap-wifi-mac.cc:
```diff
@@ -945,6 +811,13 @@ DmgApWifiMac::FrameTxOk (const WifiMacHeader &hdr)
         {
           NS_LOG_DEBUG ("DMG PCP/AP completed the transmission of the last DMG Beacon at " << Simulator::Now ());
           Time startTime = m_nextDmgBeaconDelay + GetMbifs ();
+          /* TEMPORARY FIX: BECAUSE WE CALCULATE A BTI DURATION HIGHER THAN THE ACTUAL ONE */
+          Time btiRemaining = GetBTIRemainingTime () + GetMbifs ();
+          if (btiRemaining > startTime)
+          {
+            startTime = btiRemaining;
+          }
+          /* END OF TEMPORARY FIX */
```
explain?

tommyazz replied:
Related to issue #9

mattia-lecci commented on src/wifi/model/dmg-ap-wifi-mac.cc:
```diff
@@ -1146,10 +1021,9 @@ DmgApWifiMac::EndBeaconInterval (void)
   /* Start New Beacon Interval */
-  StartBeaconInterval ();
+  /* Start beacon interval from DmgWifiScheduler */
+  //StartBeaconInterval ();
```
it's better to remove rather than comment

mattia-lecci commented on src/wifi/model/dmg-wifi-mac.h:
```diff
@@ -238,6 +238,7 @@ class DmgWifiMac : public RegularWifiMac
 
 protected:
   friend class MacLow;
+  friend class DmgWifiScheduler;
```
1. is it necessary?
2. maybe it's better to move it to DmgApWifiMac

tommyazz replied:
I need "GetMbifs" method

tommyazz committed:
Add log debug for the allocation list received at a STA

tommyazz committed:
Partially Address PR feedback from Mattia

tommyazz committed:
Fix bug, the GetTypeId of DmgWifiPhy was not called from child

tommyazz committed:
Flag to add OFDM MCSs when OFDM PHY is supported

tommyazz committed:
Not considering Extended SC MCSs for now

tommyazz committed:
Fix app start time when sending another ADDTS request to modify an alloc

tommyazz committed:
Propose fix to transmission bounded issue

tommyazz committed:
Add logs at DmgWifiScheduler and Evaluate lroom

MatteoDrago commented on src/wifi/model/dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,324 @@
+  bool m_isAddtsAccepted;                       //!< An ADDTS request received in the last DTI has been accepted.
+  bool m_isAllocationModified;                  //!< An allocation has been successfully modified.
+  bool m_isNonStatic;                           //!< A non-static allocation has been served in the last DTI.
+  bool m_isDeltsReceived;                       //!< A DELTS request has been received in the last DTI. 
```
Is it "safe" to use a class variable for a single examined ADDTS, DELTS or non-static allocation? What if we receive more than one? Is the code related still valid? We can discuss it in a call, I put it here just as a reminder (I looked also at the related code ofc)

tommyazz replied:
We forget to discuss this. 
Ideally I just need one DELTS, new or modification ADDTS request to trigger the AddBroadcastCbapAllocations method, because something changed in the addtsAllocationList. 
Having different booleans can help with identifying  what exactly happend in the last DTI.
Let's discuss offline in case

MatteoDrago replied:
But I mean, once one variable changes its value (from false to true for example) in one BI, then it stays true for the entire duration of the simulation or does it refresh at the beginning of every BI? 
My point is, let's be sure that everything is updated correctly :)

tommyazz replied:
Yes they are updated accordingly after the method AddBraodcastCbapAllocations is called.

MatteoDrago reviewed (commented):
Need further discussion on some changes. This is actually a massive work, we are doing great

tommyazz committed:
Add guard time between adjacent broadcast CBAP allocations

tommyazz committed:
Address minor feedback from PR

tommyazz committed:
Remove goto syntax in ModifyExistingAllocation

Add new ADDTS requests in the evaluation script in order to test the addition/modification of multiple allocations.

tommyazz committed:
Add simple check for the SP duration. Create new ADDTS request to reject

tommyazz committed:
Add Scheduler creation from the DmgWifiHelper

tommyazz committed:
Fix problem with inter-allocation of broadcast CBAPs

Tested the allocation of CBAPS among requested SPs.

mattia-lecci commented on scratch/evaluate_qd_channel_lroom_scenario.cc:
```diff
@@ -191,14 +192,17 @@ CongStateTrace (TcpSocketState::TcpCongState_t oldState, TcpSocketState::TcpCong
 {
+  /* Simple assert for the moment */
+  NS_ASSERT_MSG (minAllocation <= MAX_SP_BLOCK_DURATION, "Exceeding Max SP block duration");
+  NS_ASSERT_MSG (maxAllocation <= MAX_SP_BLOCK_DURATION, "Exceeding Max SP block duration");
```
i would check instead
NS_ASSERT_MSG (minAllocation < maxAllocation, "Min allocation should be < than max allocation");

MatteoDrago replied:
Agree with Mattia (but using <=) but I would leave the check on maxAllocation

mattia-lecci replied:
Yes, minAllocation <= maxAllocation is better to allow a deterministic allocation.
The conditions to check are
minAllocation <= maxAllocation <= MAX_SP_BLOCK_DURATION
thus checking minAllocation <= MAX_SP_BLOCK_DURATION is useless

mattia-lecci commented on scratch/evaluate_qd_channel_lroom_scenario.cc:
```diff
@@ -215,37 +219,35 @@ StationAssociated (Ptr<DmgStaWifiMac> staWifiMac, Mac48Address address, uint16_t
-    // staWifiMac->CreateAllocation (GetDmgTspecElement (10000, 20000));
-    // Simulator::Schedule (Seconds (1.0), &DmgStaWifiMac::CreateAllocation, staWifiMac, GetDmgTspecElement (10000, 10000));
+    staWifiMac->CreateAllocation (GetDmgTspecElement (1, true, 10000, 10000));
+    //staWifiMac->CreateAllocation (GetDmgTspecElement (2, true, 10000, 10000));
+    //staWifiMac->CreateAllocation (GetDmgTspecElement (3, true, 10000, 10000));
+    //Simulator::Schedule (Seconds (1.0), &DmgStaWifiMac::CreateAllocation, staWifiMac, GetDmgTspecElement (1, true, 30000, 30000));
+    //Simulator::Schedule (Seconds (1.0), &DmgStaWifiMac::CreateAllocation, staWifiMac, GetDmgTspecElement (3, true, 20000, 20000));
+    //Simulator::Schedule (Seconds (1.5), &DmgStaWifiMac::CreateAllocation, staWifiMac, GetDmgTspecElement (4, true, MAX_SP_BLOCK_DURATION, MAX_SP_BLOCK_DURATION));
```
avoid commenting portions of code, delete them instead.
If needed, just go back to a previous commit and either revert it or simply copy-paste it back.
If it is useful code, instead, keep it locally, committing only the relevant code.

MatteoDrago replied:
agree

mattia-lecci commented on scratch/evaluate_qd_channel_lroom_scenario.cc:
```diff
@@ -391,8 +390,14 @@ main (int argc, char *argv[])
+      //LogComponentEnable ("DmgWifiMac", LOG_LEVEL_ALL);
+      //LogComponentEnable ("EdcaTxopN", LOG_LEVEL_ALL);
+      //LogComponentEnable ("DcaTxop", LOG_LEVEL_ALL);
+      //LogComponentEnable ("MacLow", LOG_LEVEL_ALL);
```
avoid pushing commented code

mattia-lecci commented on src/wifi/helper/dmg-wifi-helper.cc:
```diff
@@ -264,6 +287,14 @@ DmgWifiHelper::Install (const SpectrumDmgWifiPhyHelper &phyHelper,
       Ptr<SpectrumDmgWifiPhy> phy = StaticCast<SpectrumDmgWifiPhy> (phyHelper.Create (node, device));
       Ptr<Codebook> codebook = m_codeBook.Create<Codebook> ();
+      Ptr<DmgApWifiMac> ap = DynamicCast<DmgApWifiMac> (mac); // safe downcast
+      if (ap)
```
maybe add a brief comment to make this operation clearer to others

tommyazz committed:
Fix creation of scheduler object if AP through helper

The creation of scheduler was missing when calling Install function with DmgWifiPhyHelper passed.
Add description.

tommyazz committed:
Evaluate an ADDTS request as soon as it is received

tommyazz committed:
Send ADDTS response to destination STA if status code success

tommyazz committed:
Send ADDTS response to destination STA in any case

According to the standard, an ADDTS response to the destination STA of the allocation is sent even if its status code is different from Success.

tommyazz committed:
Add BasicDmgWifiScheduler which implements the abstract DmgWifiScheduler

The child class BasicDmgWifiScheduler implements the pure virtual functions declared by DmgWifiScheduler. 
It provides the implementation of a basic admission control policy and DTI scheduling strategy for 80211ad.

MatteoDrago commented on scratch/evaluate_qd_channel_lroom_scenario.cc:
```diff
@@ -189,33 +191,63 @@ CongStateTrace (TcpSocketState::TcpCongState_t oldState, TcpSocketState::TcpCong
-      onoff->StartApplication ();
-      /* Connect to TCP traces */
-      if (socketType == "ns3::TcpSocketFactory")
+      appStartTime = Simulator::Now ();
```
What happens if we have multiple apps (even only two) running on the same device? Does the code remain consistent?

tommyazz replied:
No, one would need to modify the evaluation script accordingly.

MatteoDrago commented on scratch/evaluate_qd_channel_lroom_scenario.cc:
```diff
@@ -355,8 +390,15 @@ main (int argc, char *argv[])
       LogComponentEnable ("Mobility", LOG_LEVEL_ALL);
+      LogComponentEnable ("BasicDmgWifiScheduler", LOG_LEVEL_ALL);
+      LogComponentEnable ("DmgWifiScheduler", LOG_LEVEL_ALL);
+      //LogComponentEnable ("DmgApWifiMac", LOG_LEVEL_ALL);
+      //LogComponentEnable ("DmgWifiMac", LOG_LEVEL_ALL);
+      //LogComponentEnable ("EdcaTxopN", LOG_LEVEL_ALL);
+      //LogComponentEnable ("DcaTxop", LOG_LEVEL_ALL);
+      //LogComponentEnable ("MacLow", LOG_LEVEL_ALL);
```
delete if not needed or uncomment :)

mattia-lecci commented on src/wifi/helper/dmg-wifi-helper.cc:
```diff
@@ -191,6 +191,15 @@ DmgWifiHelper::Install (const DmgWifiPhyHelper &phyHelper,
       Ptr<DmgWifiMac> mac = StaticCast<DmgWifiMac> (macHelper.Create ());
       Ptr<DmgWifiPhy> phy = StaticCast<DmgWifiPhy> (phyHelper.Create (node, device));
       Ptr<Codebook> codebook = m_codeBook.Create<Codebook> ();
+      Ptr<DmgApWifiMac> ap = DynamicCast<DmgApWifiMac> (mac); // safe downcast
```
i would move it up, right after the definition of the mac variable and call it apMac, instead

mattia-lecci commented on src/wifi/helper/dmg-wifi-helper.cc:
```diff
@@ -288,6 +297,7 @@ DmgWifiHelper::Install (const SpectrumDmgWifiPhyHelper &phyHelper,
       Ptr<SpectrumDmgWifiPhy> phy = StaticCast<SpectrumDmgWifiPhy> (phyHelper.Create (node, device));
       Ptr<Codebook> codebook = m_codeBook.Create<Codebook> ();
       Ptr<DmgApWifiMac> ap = DynamicCast<DmgApWifiMac> (mac); // safe downcast
```
i would move it up, right after the definition of the mac variable and call it apMac, instead

mattia-lecci reviewed (commented):
Just as a reminder, here a few more general comments:
- [x] Have you thought about a good strategy for protected/private methods in `DmgWiFiScheduler`?
- [x] Rember that `virtual=0` methods HAVE to be implemented by child classes. There are 6 now in `DmgWifiScheduler`, could they be too many? Could they reduce the flexibility intended for this abstract class?
- [x] In the newly added classes, try to use an efficient input/output arguments strategy. Specifically, for input-only variables which are not fundamental types prefer passing it by const reference insted of value (i.e., `const T&` instead of `T`). [Here](https://www.modernescpp.com/index.php/c-core-guidelines-how-to-pass-function-parameters) you can find some guidelines on the topic

tommyazz committed:
Address PR for DmgWifiHelper

tommyazz committed:
Modify asserts and clean code in Lroom evaluation

tommyazz committed:
Add enable log components for Dmg MAC and PHY classes

tommyazz committed:
Pass arguments as const reference where possible

tommyazz committed:
Add const reference parameters for SendAddtsResponse method

tommyazz committed:
Modify with reference SetAllocationList and SendAddtsResponse
